### PR TITLE
Fix terraform blockname regex

### PR DESCRIPTION
**Pull request by dwc0011**
The regex for blockname does not allow empty provider blocks.
For example as mentioned by @lorengordon in #1353 the code snippet he provided had an empty provider block which is valid teraform. @Anteru also pointed out the issue with the first "}", at the time thought to be unrelated, however that is the reason why the snippet provided by @lorengordon was receiving the error.

The following valid terraform does not process correctly.
`provider "aws" {}`

I have updated the blockname regex to allow `{}|{`  this fixes the issue and passes all tests.  I have added a test to the snippets and ran the golden updates command on it.
`python -m pytest --update-goldens tests/snippets/terraform/test_types.txt`
I then ran `make test` and all tests passed.

Admittedly, this is an oversimplified fix and I am working on stronger regex to deal with blocknames that are not allowed to be empty.  But to get the ball rolling and ensure the tests are correct and heading down the right path I wanted to get this PR started.  

If there is any reason to put more complex or fuller terraform tests under examples, for this I just followed the instructions under contributing and added a snippets test, but am wondering terraform is at the point of needing fuller more complex test case.

Please let me know if you want me to create a separate issue or use #1353 as the issue this PR addresses.

### CODE REVIEW

**Comment by birkenfeld** [comment]
Strange, I don't understand how the lexer leaves the `blockname` state without producing an error...

**Comment by dwc0011** [comment]
> Strange, I don't understand how the lexer leaves the `blockname` state without producing an error...

Can you elaborate?  Are you saying with the fix in the regex, or before the fix in the regex?

**Comment by birkenfeld** [comment]
For both cases, the `blockname` state has no "exit" condition.

I now re-read the code, and the lexer has an implicit "return to root state" if at line end and no rules in a state matched.  This is not quite clean though, and I wonder if it should be removed.

In any case, please add a `default('#pop')` tot he state.

**Comment by dwc0011** [comment]
> For both cases, the `blockname` state has no "exit" condition.
> 
> I now re-read the code, and the lexer has an implicit "return to root state" if at line end and no rules in a state matched. This is not quite clean though, and I wonder if it should be removed.
> 
> In any case, please add a `default('#pop')` tot he state.

Ok, I didn't look at whether the lexer was correct, once I noticed the regex I adjusted that and then ran the tests.  I will look at it closer and add the default('#pop')  and test again.

**Comment by lorengordon** [comment]
@birkenfeld Anything else needed here?

File: pygments/lexers/configs.py [hunk 1]
```diff
@@ -726,8 +726,9 @@ def heredoc_callback(self, match, ctx):
         'blockname': [
             # e.g. resource "aws_security_group" "allow_tls" {
             # e.g. backend "consul" {
-            (r'(\s*)("[0-9a-zA-Z-_]+")?(\s*)("[0-9a-zA-Z-_]+")(\s+)(\{)',
+            (r'(\s*)("[0-9a-zA-Z-_]+")?(\s*)("[0-9a-zA-Z-_]+")(\s+)(\{\}|\{)',
```
[the remark this answers is not in the document]
Comment (dwc0011) [reply pygments/lexers/configs.py]: Removed (\s+)(\{) fom the regex

File: pygments/lexers/configs.py [hunk 2]
```diff
@@ -726,8 +726,28 @@ def heredoc_callback(self, match, ctx):
...
+                    Whitespace,
+                    Punctuation,
+                ),
+            ),
+            default("#pop"),
+=======
+            (r'(\s*)("[0-9a-zA-Z-_]+")?(\s*)("[0-9a-zA-Z-_]+")',
+	    bygroups(Whitespace, Name.Class, Whitespace, Name.Variable, Whitespace, Punctuation)),
+            default('#pop'),
+>>>>>>> Stashed changes
```
Comment (Jean-Abou-Samra) [inline pygments/lexers/configs.py:750]: You have unresolved merge conflicts.

**Comment by Jean-Abou-Samra** [comment]
Unfortunately, this PR still has merge conflicts with the master branch.

**Comment by dwc0011** [comment]
> Unfortunately, this PR still has merge conflicts with the master branch.

I assumed we would deal with merge issue with master once everything was reviewed and agreed it is good.

I will address now.

**Comment by Jean-Abou-Samra** [comment]
> I assumed we would deal with merge issue with master once everything was reviewed and agreed it is good.

Well, it happens that updates to the merge target completely change the game as to the way a change should be done. Plus, for such a small change, it shouldn't be a burden to rebase.

File: pygments/lexers/configs.py [hunk 3]
```diff
@@ -729,6 +729,13 @@ def heredoc_callback(self, match, ctx):
                 heredoc_callback,
             )
         ],
+        'blockname': [
+            # e.g. resource "aws_security_group" "allow_tls" {
+            # e.g. backend "consul" {
+            (r'(\s*)("[0-9a-zA-Z-_]+")?(\s*)("[0-9a-zA-Z-_]+")',
+             bygroups(Whitespace, Name.Class, Whitespace, Name.Variable, Whitespace, Punctuation)),
```
Comment (Jean-Abou-Samra) [inline pygments/lexers/configs.py:736]: Fails with `regexlint` in the CI, it complains that there are only 4 groups for 6 token types in bygroups(). I tend to agree :-)

**Comment by dwc0011** [comment]
> I was surprised to see a fresh and unused state in the diff -- apparently the same issue has already been fixed by #2097 in the meantime?

I am glad theirs got merged, this one has been open since December and never was addressed until recently, such is the life of open source projects.  We will test the latest build and see if it fixes the issues we were addressed with the other fix.

**Comment by lorengordon** [comment]
@Jean-Abou-Samra Any idea when the 2.12 release will be, so we can see if #2097 does indeed fix this issue?

**Comment by Jean-Abou-Samra** [comment]
@lorengordon You'd have to ask @Anteru. However, there is no need to wait on the release. Just do

```
pip install --user https://github.com/pygments/pygments/archive/master.zip
```

and (provided I've got the command right) that should give you an installation with current master. For your information, this is the token output I get for the example from https://github.com/pygments/pygments/issues/1353#issuecomment-862414618:

```
Token.Keyword.Reserved	'provider'
Token.Text.Whitespace	' '
Token.Name.Variable	'"aws"'
Token.Text.Whitespace	' '
Token.Punctuation	'{'
Token.Punctuation	'}'
Token.Text.Whitespace	'\n'
Token.Text.Whitespace	'\n'
Token.Keyword.Reserved	'module'
Token.Text.Whitespace	' '
Token.Name.Variable	'"test-lx-instance"'
Token.Text.Whitespace	' '
Token.Punctuation	'{'
Token.Text.Whitespace	'\n'
Token.Text.Whitespace	'  '
Token.Name.Attribute	'source'
Token.Text.Whitespace	' '
Token.Operator	'='
Token.Text.Whitespace	' '
Token.Literal.String.Double	'"git::https://github.com/plus3it/terraform-aws-watchmaker//modules/lx-instance/"'
Token.Text.Whitespace	'\n'
Token.Text.Whitespace	'\n'
Token.Text.Whitespace	'  '
Token.Name.Attribute	'Name'
Token.Text.Whitespace	'      '
Token.Operator	'='
Token.Text.Whitespace	' '
Token.Literal.String.Double	'"tf-watchmaker-lx-autoscale"'
Token.Text.Whitespace	'\n'
Token.Text.Whitespace	'  '
Token.Name.Attribute	'AmiId'
Token.Text.Whitespace	'     '
Token.Operator	'='
Token.Text.Whitespace	' '
Token.Literal.String.Double	'"__AMIID__"'
Token.Text.Whitespace	'\n'
Token.Text.Whitespace	'  '
Token.Name.Attribute	'AmiDistro'
Token.Text.Whitespace	' '
Token.Operator	'='
Token.Text.Whitespace	' '
Token.Literal.String.Double	'"__AMIDISTRO__"'
Token.Text.Whitespace	'\n'
Token.Punctuation	'}'
Token.Text.Whitespace	'\n'
```

**Comment by lorengordon** [comment]
Appreciate that, thank you!

**Comment by Jean-Abou-Samra** [comment]
You're welcome. Do you confirm that the bug went away?

**Pull request closed without merging**